### ISSUE TITLE
Services Config Files

### ISSUE THREAD

**Issue by davidbarratt** [young-steveo/bottlejs#57]
Any plans to support services config files?
http://symfony.com/doc/current/components/dependency_injection/introduction.html#setting-up-the-container-with-configuration-files

It would be really awesome if you could define your services in (json?) files rather than doing all of it by code.

**Comment by young-steveo** [comment]
That's interesting.  One downside I see is that your services would need to be globally resolvable, so that the configuration can find it and instantiate it.  In PHP this is fine, because we have Namespaces.  I'm not sure how to solve that problem without cluttering up the global scope in JS.

**Comment by young-steveo** [comment]
To be clear:  In node you could specify a path to the module, but I'm not sure how to elegantly handle the problem in a browser.

**Comment by davidbarratt** [comment]
so... I mean the only ways I know of to handle non-globals would be either to use `require()` (and let the user use [browserify](http://browserify.org/) or the like) or to use `import` (with [babel](http://babeljs.io/docs/plugins/). 

so you could I guess make a config option to either use `require()` or `import`. Although, I don't think `import` can be called anywhere other than at the top of the file (before any code execution). 

So basically, to use the config on the browser, you'd have to use another library until the browser natively supports `import`

**Comment by davidbarratt** [comment]
so you kind of can dynamically load ES2015 modules:
http://babeljs.io/docs/learn-es2015/#module-loaders

so that's how I would do it. :) Obviously the paths are going to be relative to the web root.

**Comment by young-steveo** [comment]
I'm going to put some more thought into it.  Meanwhile, I found a project that is close to what you're looking for: https://github.com/rouflak/barmanjs  Albeit, it would be nice to configure everything with a pure JSON file (that project requires imports and putting the module constructors into the configuration object itself).

**Comment by bingomanatee** [comment]
I actually would put a vote to the contrary. Given that bottle is not tied right now to a particular ecosystem, a configuration based approach would impose a specific type of getting the resource before it is registered. 

If you wanted to create a sort of "Metabottle" (Case.js) solution to translate that injection into a "bottler" you could but I'd think its better as a sibling project than as a direct feature of bottle. As it is, bottle is usable across any sort of implementation system or context, and enabling a feature that diverges from that seems like a step in the wrong direction.

**Comment by young-steveo** [comment]
@bingomanatee If implemented, config file injection would be completely optional.  All existing bottle functionality would remain unchanged.

**Comment by bingomanatee** [comment]
I understand.

Also, thanks for making this. I'm [Writing](http://wonderlandlabs.com/blog/react/having_it_all) on this in my blog and find it very useful. Right now, in Node for testing, very soon in React.js.

My concern is, I don't want the fact that Bottle.js is universally applicable to become obscured by example use cases tied to a particular style of resource loading.

**Comment by young-steveo** [comment]
You're welcome, @bingomanatee.  👍 

I would consider a sibling project for config file injection if the amount of code to make it work got too long (one of Bottle's goals is to remain under 5k when minified and gzipped).

I read your draft; I like the fact that you're including a real world example.  Thanks for using bottle!

**Comment by young-steveo** [comment]
A proposal:

When in a node environment, a JSON configuration file will define a path to `require` each service.  In a browser environment a global variable "namespace" must be defined.  Bottle will use `window.YourNameSpace.YourService` to lookup the service definitions, so in order to use file based configuration, you would need to attach all of your services to this global object.

Three new methods will support building the container with configuration objects:
```js
Bottle.build({ /* config */ }); // returns a configured bottle instance.
bottleInstance.build({ /* config */ }); // configures and returns the bottleInstance
bottleInstance.container.$build({ /* config */ }); // configures and returns the container
```


```js
{
  /**
   * required only in the browser.
   * The global property in which all services are located.  i.e. window.someGlobal
   */
  "namespace" : "someGlobal",

  /**
   * optional - When using the global, static `Bottle.config` method,
   * this will give the instance a name, so that future calls to `Bottle.pop('name')`
   * will properly resolve.
   */
  "name" : "MyBottle",

  /**
   * required - an array of service configurations
   */
  "services" : [
    {
        /**
         * required - the name used to register the object
         */
        "name" : "Beer",

        /**
         * optional - the method used to register the object. Defaults to 'service'.
         * Valid types are: 'service', 'factory', 'provider', 'value', 'constant'
         */
        "type" : "service",

        /**
         * optional - an array of names of dependencies to inject into the constructor
         * if `type` was 'service'
         */
        "inject" : [ "Water", "Hops", "Barley" ],

        /**
         * required in node - the path to load the module, relative to the bottle.config() call.
         * optional in the browser - the property on the global object to lookup the service
         *               for registration.  Defaults to the "name" of the service
         */
        "path" : "../lib/Beer", // in node
        "path" : "someOtherProperty" // in browser
    }
  ]
}
```

**Comment by young-steveo** [comment]
After about--checks watch--four years of thinking about this change, I'm going to close the ticket.  It seems trivial enough for a user that needs this kind of thing to write a script to do it, and including it in bottlejs would just add unnecessary  kilobytes to the browser builds (something bottle tries not to do).
